migration/src/bin/migrate_to.rs

```diff
@@ -0,0 +1,393 @@
+    sea_orm::{ConnectOptions, Database, DatabaseConnection, EntityTrait, QueryOrder},
+    MigratorTrait,
+};
+use sea_orm_migration::{prelude::*, seaql_migrations};
+use tracing::{info, Level};
+
+#[derive(Copy, Clone, ValueEnum, Debug)]
+enum Direction {
+    Up,
+    Down,
+}
+
+/// Wraps the SeaORM migration library to provide additional features. Lets
+/// you bring migrations up or down to a particular version. Allows dry-run
+/// of migrations.
+#[derive(Parser, Debug)]
+#[command(about, version)]
+struct Args {
+    #[arg(value_enum)]
+    direction: Direction,
+    #[arg(value_parser = available_migrations())]
+    target_version: String,
+    #[arg(short, long)]
+    dry_run: bool,
+    #[arg(short = 'u', long, env = "DATABASE_URL")]
+    database_url: String,
+}
+
+#[async_std::main]
+async fn main() -> Result<(), Error> {
+    let args = Args::parse();
+    tracing_subscriber::fmt()
+        .with_max_level(Level::DEBUG)
+        .init();
+
+    let db = Database::connect(ConnectOptions::new(args.database_url)).await?;
+    migrate_to::<migration::Migrator>(&db, args.dry_run, args.direction, &args.target_version)
+        .await?;
+    Ok(())
+}
+
+async fn migrate_to<M: MigratorTrait>(
+    db: &DatabaseConnection,
+    dry_run: bool,
+    dir: Direction,
+    target: &str,
+) -> Result<(), Error> {
+    let latest_index = match latest_applied_migration(db).await {
+        Ok(m) => migration_index::<M>(&m).ok_or(Error::DBMigrationNotFound)?,
+        Err(_) => -1, // Empty database.
+    };
+    let target_index =
+        migration_index::<M>(target).ok_or(Error::MigrationNotFound(target.to_string()))?;
+    if target_index == latest_index {
+        info!("no action taken, already at desired version");
+        return Ok(());
+    }
+
+    match dir {
+        Direction::Up => match target_index - latest_index {
+            num_migrations if num_migrations > 0 => {
+                info!("executing {num_migrations} migration(s) to {target}");
+                if !dry_run {
+                    M::up(db, Some(num_migrations as u32))
+                        .await
+                        .map_err(Error::from)?
+                }
+                Ok(())
+            }
+            _ => Err(Error::VersionTooOld(target.to_string())),
+        },
+        Direction::Down => match latest_index - target_index {
+            num_migrations if num_migrations > 0 => {
+                info!("executing {num_migrations} migration(s) to {target}");
+                if !dry_run {
+                    M::down(db, Some(num_migrations as u32))
+                        .await
+                        .map_err(Error::from)?
+                }
+                Ok(())
+            }
+            _ => Err(Error::VersionTooNew(target.to_string())),
+        },
+    }
+}
+
+fn migration_index<M: MigratorTrait>(version: &str) -> Option<i64> {
+    M::migrations()
+        .iter()
+        .position(|m| m.name() == version)
+        .map(|i| i as i64)
```

Comment:
I'm not crazy about these casts. `Iterator::position` gives you `usize`, which you then cast to `i64` here. Later, you do a little math with them but finally pass them to `MigratorTrait::up` or `::down`, which takes `Option<u32>`. I think you could use `u32::try_from::<usize>` to safely handle the conversion, and then make this method return `Result<u32, Error>`, especially since the existing callsites of `migration_index` already have to handle the case where it returns `None` and do so by converting into `Result<i64, Error>` anyway.

Actually, I think we could get away with unwrapping `u32::try_from::<usize>` and not returning an error here, as this tool is unlikely to ever be used anywhere that `usize` is less than 32 bits!

Reply:
I was able to eliminate the unsafe cast. It did lead to some rather extensive changes, most notably I went ahead and broke out `migrate_to` into `migrate_down` and `migrate_up`, because the combined approach was turning out to be unreadable.